HLTrigger/Configuration/python/customizeHLTforCMSSW.py

```diff
@@ -314,6 +314,82 @@ def customizeHLTfor45063(process):
+            if not hasattr(prod.iCfgCandConnector, "ptErrorSecondary"):
+                prod.iCfgCandConnector.ptErrorSecondary = cms.double(0)
+            if not hasattr(prod.iCfgCandConnector, "nuclCalibFactors"):
+                prod.iCfgCandConnector.nuclCalibFactors = cms.vdouble()
+
+    # Fix types for gathering and seeding thresholds, and recHitEnergyNorm
+    for prod in producers_by_type(process, 'PFClusterProducer'):
+        if hasattr(prod, "initialClusteringStep") and hasattr(prod.initialClusteringStep, "thresholdsByDetector"):
+            if hasattr(prod.initialClusteringStep.thresholdsByDetector, "gatheringThreshold") and isinstance(prod.initialClusteringStep.thresholdsByDetector.gatheringThreshold, cms.vdouble):
+                prod.initialClusteringStep.thresholdsByDetector.gatheringThresholds = prod.initialClusteringStep.thresholdsByDetector.gatheringThreshold
+                delattr(prod.initialClusteringStep.thresholdsByDetector, "gatheringThreshold")
+            if hasattr(prod.initialClusteringStep.thresholdsByDetector, "gatheringThresholdPt") and isinstance(prod.initialClusteringStep.thresholdsByDetector.gatheringThresholdPt, cms.vdouble):
+                prod.initialClusteringStep.thresholdsByDetector.gatheringThresholdsPt = prod.initialClusteringStep.thresholdsByDetector.gatheringThresholdPt
+                delattr(prod.initialClusteringStep.thresholdsByDetector, "gatheringThresholdPt")
+        if hasattr(prod, "seedFinder") and hasattr(prod.seedFinder, "thresholdsByDetector"):
+            if hasattr(prod.seedFinder.thresholdsByDetector, "seedingThreshold") and isinstance(prod.seedFinder.thresholdsByDetector.seedingThreshold, cms.vdouble):
+                prod.seedFinder.thresholdsByDetector.seedingThresholds = prod.seedFinder.thresholdsByDetector.seedingThreshold
+                delattr(prod.seedFinder.thresholdsByDetector, "seedingThreshold")
+            if hasattr(prod.seedFinder.thresholdsByDetector, "seedingThresholdPt") and isinstance(prod.seedFinder.thresholdsByDetector.seedingThresholdPt, cms.vdouble):
+                prod.seedFinder.thresholdsByDetector.seedingThresholdsPt = prod.seedFinder.thresholdsByDetector.seedingThresholdPt
+                delattr(prod.seedFinder.thresholdsByDetector, "seedingThresholdPt")
+        for n in ["positionCalc", "allCellsPositionCalc"]:
+            if hasattr(prod, "pfClusterBuilder") and hasattr(prod.pfClusterBuilder, n) and hasattr(getattr(prod.pfClusterBuilder, n), "recHitEnergyNorms") and hasattr(getattr(prod.pfClusterBuilder, n).recHitEnergyNorms, "recHitEnergyNorm") and isinstance(getattr(prod.pfClusterBuilder, n).recHitEnergyNorms.recHitEnergyNorm, cms.vdouble):
+                getattr(prod.pfClusterBuilder, n).recHitEnergyNorms.recHitEnergyNorms = getattr(prod.pfClusterBuilder, n).recHitEnergyNorms.recHitEnergyNorm
+                delattr(getattr(prod.pfClusterBuilder, n).recHitEnergyNorms, "recHitEnergyNorm")
+
+    return process
```

Comment:
Hmm, why adding all these new parameters here - when there is the fillDescriptions methods which take care of providing them if they are not in the py config?

Reply:
Are the values to be used different?

Reply:
Because when I tested them, if there were not present in the menu, it returned an error of missing parameter. So this checks if the parameter is present, and if it is not, then adds a default value to the HLT module

Reply:
Or are you referring to the recHitEnergyNorms , seedingThresholds and gatheringThresholds parameters?
If yes, it is because there were conflicts of two different types (double and std::vector) used with exactly the same parameter name (not allowed in fillDescription, so I created two sets of parameters:

seedingThresholds , gatheringThresholds and recHitEnergyNorms of type std::vector
seedingThreshold , gatheringThreshold and recHitEnergyNorm of type double (keeping the original names)

So the python macro also checks for the parameters which are type std::vector with names (seedingThreshold , gatheringThreshold and recHitEnergyNorm) and effectively renames them to the new names (seedingThresholds , gatheringThresholds and recHitEnergyNorms)

Reply:
Not sure I follow. The fillDescriptions have a default added

Reply:
It it does the framework should not complain anni missing parameters

Reply:
> It it does the framework should not complain about missing parameters

Reply:
Well, the whole point of fillDescription parameters WITH values(!) is to use those in case the parameters are missing from the py config. So if the values in fillDescriptions are those you add also in the customisation routine (eg, empty VPSets()), then they can and should be removed from the customisation routine, as fillDescriptions is sufficient then.

Reply:
ok, then probably I am missing something since I still get the error. I will check and see if I can remove the stuff I added to the customizeforCMSSW

Reply:
ok, I think the issue is probably related to them been declared optional. I will try to change them from optional to normal parameters

Reply:
see also https://github.com/cms-sw/cmssw/pull/45212#discussion_r1641748898

Reply:
is there a way to define which parameters should a parameter set have but have as default an empty parameter set?

Reply:
So when I replaced the exists for checking the emptiness of the parameter sets, it prompted the situation where a parameter set would then be by default empty but could also have parameters, so I made the parameters of those psets optional. However the moment those psets are not empty anymore, then some of the parameters are read and it triggers the missing parameter error when they are not present (since they were declared as optional).
I will check how to deal with this cases (probably avoid checking the the psets is empty and instead use another criteria to replace the previos exists conditions)

Reply:
Fixed it in c85415bc65ac6663b9a4d267d545c6402d82a540 , such that now all the parameters I added in the fillDescription are no longer optional. Also fixed the customizeForCMSSW.py file such that now it only deals with the ranaming of the double vector parameters

Reply:
I managed to undo the renaming of the double vector parameters in 9fee0bbed1f2099220fa99c13d22fcc9c84e62fe , so now I don't customize the menu anymore, and this change helped reduce the number of changed files by 10